File: maskql/schemas/rule.py
from typing import Optional
from sqlmodel import SQLModel

class RuleCreate(SQLModel):
    schema_name: Optional[str] = None
    table_name: Optional[str] = None
    column_name: Optional[str] = None
    allow: bool
    effect: Optional[str] = None
    catalog_id: Optional[int] = None
    catalog: Optional[str] = None # Rule can be created with catalog name
    user_id: int

class RulePatch(SQLModel):
    schema_name: Optional[str] = None
    table_name: Optional[str] = None
    column_name: Optional[str] = None
    allow: Optional[bool] = None
    effect: Optional[str] = None
    catalog_id: Optional[int] = None
    user_id: Optional[int] = None

class RuleRead(SQLModel):
    id: int
    schema_name: Optional[str] = None
    table_name: Optional[str] = None
    column_name: Optional[str] = None
    allow: bool
    effect: Optional[str] = None
    catalog_id: int
    user_id: int

File: frontend/src/types/rule.ts
import { BaseResource } from "./base";

export interface Rule {
    id: number;
    schema_name: string;
    table_name: string;
    column_name: string;
    allow: boolean;
    effect: string; 
    catalog_id: number;
    user_id: number;
}

export type RuleCreate = Omit<Rule, "id">;
export type RuleUpdate = Partial<Omit<Rule, "id">>;

export class Rules extends BaseResource<Rule, RuleCreate, RuleUpdate> {
    constructor() {
        super("/rules");
    }

    async listByCatalog(catalogId: number): Promise<Rule[]> {
        return this.list({ catalog_id: catalogId });
    }
    async listByUser(userId: number): Promise<Rule[]> {
        return this.list({ user_id: userId });
    }
}

export const RuleAPI = new Rules();
